packages/wdio-browserstack-service/src/testOps/listener.ts

```diff
@@ -0,0 +1,227 @@
+import UsageStats from './usageStats.js'
+import type FeatureStats from './featureStats.js'
+import RequestQueueHandler from '../request-handler.js'
+import type { LogData, ScreenshotLog, TestData, UploadType } from '../types.js'
+import { batchAndPostEvents, sleep } from '../util.js'
+import {
+    DATA_BATCH_ENDPOINT,
+    DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS,
+    DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS,
+    LOG_KIND_USAGE_MAP
+} from '../constants.js'
+import { sendScreenshots } from './requestUtils.js'
+import { BStackLogger } from '../bstackLogger.js'
+import * as util from 'node:util'
+import fs from 'node:fs'
+import path from 'node:path'
+
+class Listener {
+    private static instance: Listener
+    private readonly usageStats: UsageStats = UsageStats.getInstance()
+    private readonly testStartedStats: FeatureStats = this.usageStats.testStartedStats
+    private readonly testFinishedStats: FeatureStats = this.usageStats.testFinishedStats
+    private readonly hookStartedStats: FeatureStats = this.usageStats.hookStartedStats
+    private readonly hookFinishedStats: FeatureStats = this.usageStats.hookFinishedStats
+    private readonly cbtSessionStats: FeatureStats = this.usageStats.cbtSessionStats
+    private readonly logEvents: FeatureStats = this.usageStats.logStats
+    private requestBatcher?: RequestQueueHandler
+
+    private constructor() {
+    }
+
+    public static getInstance(): Listener {
+        if (!Listener.instance) {
+            Listener.instance = new Listener()
+        }
+        return Listener.instance
+    }
+
+    public async onWorkerEnd() {
+        await this.uploadPending()
+        await this.teardown()
+        this.saveData(this.usageStats.getDataToSave())
+    }
+
+    async uploadPending(waitTimeout = DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS, waitInterval = DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS): Promise<unknown> {
+        // @ts-ignore
+        if (this.requestBatcher?.pendingUploads <= 0 || waitTimeout <= 0) { // TODO: remove ts-ignore
+            return
+        }
+
+        await sleep(waitInterval)
+        return this.uploadPending(waitTimeout - waitInterval)
+    }
+
+    async teardown() {
```

Comment:
Add void as a type for the fn

Reply:
Generally typescript doesn't mandate specifying data types for functions, as it implicitly computes the returned variables. So we generally doesn't need to specify return type for functions

Also in this case, we need to declare the return type as Promise<void>, which not much readbable, so avoided